causal_testing/testing/estimators.py

```diff
@@ -599,3 +600,238 @@ def estimate_coefficient(self, bootstrap_size=100) -> tuple[pd.Series, list[pd.S
+        self.treatment_strategy = treatment_strategy
+        self.outcome = outcome
+        self.fault_column = fault_column
+        self.timesteps_per_intervention = timesteps_per_intervention
+        self.fit_bl_switch_formula = fit_bl_switch_formula
+        self.fit_bltd_switch_formula = fit_bltd_switch_formula
+        self.eligibility = eligibility
+        self.df = df
+        self.preprocess_data()
+
+    def add_modelling_assumptions(self):
+        self.modelling_assumptions.append("The variables in the data vary over time.")
+
+    def setup_xo_t_do(self, strategy_assigned: list, strategy_followed: list, eligible: pd.Series):
+        """
+        Return a binary sequence with each bit representing whether the current
+        index is the time point at which the individual diverted from the
+        assigned treatment strategy (and thus should be censored).
+
+        :param strategy_assigned - the assigned treatment strategy
+        :param strategy_followed - the strategy followed by the individual
+        :param eligible - binary sequence represnting the eligibility of the individual at each time step
+        """
+        strategy_assigned = [1] + strategy_assigned + [1]
+        strategy_followed = [1] + strategy_followed + [1]
+
+        mask = (
+            pd.Series(strategy_assigned, index=eligible.index) != pd.Series(strategy_followed, index=eligible.index)
+        ).astype("boolean")
+        mask = mask | ~eligible
+        mask.reset_index(inplace=True, drop=True)
+        false = mask.loc[mask]
+        if false.empty:
+            return np.zeros(len(mask))
+        mask = (mask * 1).tolist()
+        cutoff = false.index[0] + 1
+        return mask[:cutoff] + ([None] * (len(mask) - cutoff))
+
+    def setup_fault_t_do(self, individual: pd.DataFrame):
+        """
+        Return a binary sequence with each bit representing whether the current
+        index is the time point at which the event of interest (i.e. a fault)
+        occurred.
+        """
+        fault = individual[~individual[self.fault_column]]
+        fault_t_do = pd.Series(np.zeros(len(individual)), index=individual.index)
+
+        if not fault.empty:
+            fault_time = individual["time"].loc[fault.index[0]]
+            # Ceiling to nearest observation point
+            fault_time = ceil(fault_time / self.timesteps_per_intervention) * self.timesteps_per_intervention
+            # Set the correct observation point to be the fault time of doing (fault_t_do)
+            observations = individual.loc[
+                (individual["time"] % self.timesteps_per_intervention == 0) & (individual["time"] < fault_time)
+            ]
+            if not observations.empty:
+                fault_t_do.loc[observations.index[0]] = 1
+        assert sum(fault_t_do) <= 1, f"Multiple fault times for\n{individual}"
+
+        return pd.DataFrame({"fault_t_do": fault_t_do})
+
+    def setup_fault_time(self, individual, perturbation=-0.001):
```

Comment:
The typing for `individual` is missing. I only noticed as I wanted to see what kind of object you can invert with ~ and use `len` on

Reply:
Fixed. It's a pandas dataframe. You invert a boolean Series object with ~.